src/test/java/com/google/cloud/spanner/jdbc/PgNumericResultSetTest.java

```diff
@@ -0,0 +1,801 @@
+
+    try (Statement statement = connection.createStatement();
+        ResultSet resultSet = statement.executeQuery(QUERY)) {
+      final ResultSetMatcher<String> matcher =
+          resultSetMatcherFrom(resultSet, ResultSet::getNString, ResultSet::getNString);
+
+      matcher.nextAndAssertEquals("0");
+      matcher.nextAndAssertEquals("1");
+      matcher.nextAndAssertEquals("1.23");
+      matcher.nextAndAssertEquals(maxScale);
+      matcher.nextAndAssertEquals(maxPrecision);
+      matcher.nextAndAssertEquals("NaN");
+      matcher.nextAndAssertEquals(null);
+    }
+  }
+
+  @Test
+  public void testGetBoolean() throws Exception {
+    mockScalarResults("0", null, "1", "NaN", "1.00");
+
+    try (Statement statement = connection.createStatement();
+        ResultSet resultSet = statement.executeQuery(QUERY)) {
+
+      final ResultSetMatcher<Boolean> matcher =
+          resultSetMatcherFrom(resultSet, ResultSet::getBoolean, ResultSet::getBoolean);
+
+      // 0 == false
+      matcher.nextAndAssertEquals(false);
+      // NULL == false
+      matcher.nextAndAssertEquals(false);
+      // anything else == true
+      matcher.nextAndAssertEquals(true); // "1" == true
+      matcher.nextAndAssertEquals(true); // "Nan" == true
+      matcher.nextAndAssertEquals(true); // "1.00" == true
+    }
+  }
+
+  @Test
+  public void testGetByte() throws Exception {
+    final String minValue = Byte.MIN_VALUE + "";
+    final String underflow = BigDecimal.valueOf(Byte.MIN_VALUE).subtract(BigDecimal.ONE).toString();
+    final String maxValue = Byte.MAX_VALUE + "";
+    final String overflow = BigDecimal.valueOf(Byte.MAX_VALUE).add(BigDecimal.ONE).toString();
+
+    mockScalarResults(minValue, maxValue, "1.23", null, "NaN", underflow, overflow);
+
+    try (Statement statement = connection.createStatement();
+        ResultSet resultSet = statement.executeQuery(QUERY)) {
+
+      final ResultSetMatcher<Byte> matcher =
+          resultSetMatcherFrom(resultSet, ResultSet::getByte, ResultSet::getByte);
+
+      matcher.nextAndAssertEquals(Byte.MIN_VALUE);
+      matcher.nextAndAssertEquals(Byte.MAX_VALUE);
+      matcher.nextAndAssertEquals((byte) 1);
+      // NULL == 0
+      matcher.nextAndAssertEquals((byte) 0);
+      matcher.nextAndAssertError(JdbcSqlExceptionImpl.class, "NaN is not a valid number");
+      matcher.nextAndAssertError(
+          JdbcSqlExceptionImpl.class, "Value out of range for byte: " + underflow);
+      matcher.nextAndAssertError(
+          JdbcSqlExceptionImpl.class, "Value out of range for byte: " + overflow);
+    }
+  }
+
+  @Test
+  public void testGetShort() throws Exception {
+    final String minValue = Short.MIN_VALUE + "";
+    final String underflow =
+        BigDecimal.valueOf(Short.MIN_VALUE).subtract(BigDecimal.ONE).toString();
+    final String maxValue = Short.MAX_VALUE + "";
+    final String overflow = BigDecimal.valueOf(Short.MAX_VALUE).add(BigDecimal.ONE).toString();
```

Comment:
I'm curious, why do we use `BigDecimal` instead of `int`/`long` for calculating overflow and underflow of Short?